Optimizer crash with Labeled+Return of inline class, discarded in statement position

When the following conditions all happen together:

* A `Labeled` block has a `Return` node returning an instance of an `@inline class`
* The optimizer can fold the result to that single `Return` (and get rid of other possible `Return`s)
* The `Labeled` is used in a position that causes it to be `pretransform`ed (passed as argument to a method, assigned to a `val`, etc.)
* But then it is deemed dead code (because the result is unused, possibility because the `val` is unused)
* And that happens as the last statement in a `Block`

Then, we get a crash of the optimizer with the message:

> `java.lang.AssertionError`: assertion failed: Cannot create a `Tree` with record type `RecordType(List())`

<details>

```
[error] org.scalajs.linker.frontend.optimizer.OptimizerCore$OptimizeException: The Scala.js optimizer crashed while optimizing helloworld.HelloWorld$.testMinimized;V: java.lang.AssertionError: assertion failed: Cannot create a Tree with record type RecordType(List())
[error] Methods attempted to inline:
[error] * helloworld.HelloWorld$.makeBug;I;Lhelloworld.Bug
[error]
[error]         at org.scalajs.linker.frontend.optimizer.OptimizerCore.optimize(OptimizerCore.scala:178)
[error]         at org.scalajs.linker.frontend.optimizer.IncOptimizer$MethodImpl.doProcess(IncOptimizer.scala:1688)
[error]         at org.scalajs.linker.frontend.optimizer.IncOptimizer$MethodImpl.doProcess(IncOptimizer.scala:1626)
[error]         at org.scalajs.linker.frontend.optimizer.IncOptimizer$Processable.process(IncOptimizer.scala:1561)
[error]         at org.scalajs.linker.frontend.optimizer.IncOptimizer.$anonfun$processAllTaggedMethods$3(IncOptimizer.scala:411)
[error]         at org.scalajs.linker.frontend.optimizer.IncOptimizer.$anonfun$processAllTaggedMethods$3$adapted(IncOptimizer.scala:411)
[error]         at scala.collection.parallel.mutable.ParArray$ParArrayIterator.foreach(ParArray.scala:145)
[error]         at scala.collection.parallel.ParIterableLike$Foreach.leaf(ParIterableLike.scala:982)
[error]         at scala.collection.parallel.Task.$anonfun$tryLeaf$1(Tasks.scala:53)
[error]         at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
[error]         at scala.util.control.Breaks$$anon$1.catchBreak(Breaks.scala:67)
[error]         at scala.collection.parallel.Task.tryLeaf(Tasks.scala:56)
[error]         at scala.collection.parallel.Task.tryLeaf$(Tasks.scala:50)
[error]         at scala.collection.parallel.ParIterableLike$Foreach.tryLeaf(ParIterableLike.scala:979)
[error]         at scala.collection.parallel.AdaptiveWorkStealingTasks$WrappedTask.internal(Tasks.scala:170)
[error]         at scala.collection.parallel.AdaptiveWorkStealingTasks$WrappedTask.internal$(Tasks.scala:157)
[error]         at scala.collection.parallel.AdaptiveWorkStealingForkJoinTasks$WrappedTask.internal(Tasks.scala:440)
[error]         at scala.collection.parallel.AdaptiveWorkStealingTasks$WrappedTask.compute(Tasks.scala:150)
[error]         at scala.collection.parallel.AdaptiveWorkStealingTasks$WrappedTask.compute$(Tasks.scala:149)
[error]         at scala.collection.parallel.AdaptiveWorkStealingForkJoinTasks$WrappedTask.compute(Tasks.scala:440)
[error]         at java.base/java.util.concurrent.RecursiveAction.exec(RecursiveAction.java:194)
[error]         at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:387)
[error]         at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.tryRemoveAndExec(ForkJoinPool.java:1351)
[error]         at java.base/java.util.concurrent.ForkJoinTask.awaitDone(ForkJoinTask.java:422)
[error]         at java.base/java.util.concurrent.ForkJoinTask.join(ForkJoinTask.java:651)
[error]         at scala.collection.parallel.ForkJoinTasks$WrappedTask.sync(Tasks.scala:379)
[error]         at scala.collection.parallel.ForkJoinTasks$WrappedTask.sync$(Tasks.scala:379)
[error]         at scala.collection.parallel.AdaptiveWorkStealingForkJoinTasks$WrappedTask.sync(Tasks.scala:440)
[error]         at scala.collection.parallel.ForkJoinTasks.executeAndWaitResult(Tasks.scala:423)
[error]         at scala.collection.parallel.ForkJoinTasks.executeAndWaitResult$(Tasks.scala:416)
[error]         at scala.collection.parallel.ForkJoinTaskSupport.executeAndWaitResult(TaskSupport.scala:60)
[error]         at scala.collection.parallel.ExecutionContextTasks.executeAndWaitResult(Tasks.scala:555)
[error]         at scala.collection.parallel.ExecutionContextTasks.executeAndWaitResult$(Tasks.scala:555)
[error]         at scala.collection.parallel.ExecutionContextTaskSupport.executeAndWaitResult(TaskSupport.scala:84)
[error]         at scala.collection.parallel.ParIterableLike.foreach(ParIterableLike.scala:465)
[error]         at scala.collection.parallel.ParIterableLike.foreach$(ParIterableLike.scala:464)
[error]         at scala.collection.parallel.mutable.ParArray.foreach(ParArray.scala:58)
[error]         at org.scalajs.linker.frontend.optimizer.ParCollOps$.foreach(ParCollOps.scala:50)
[error]         at org.scalajs.linker.frontend.optimizer.ParCollOps$.foreach(ParCollOps.scala:22)
[error]         at org.scalajs.linker.frontend.optimizer.IncOptimizer.processAllTaggedMethods(IncOptimizer.scala:411)
[error]         at org.scalajs.linker.frontend.optimizer.IncOptimizer.$anonfun$update$4(IncOptimizer.scala:128)
[error]         at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
[error]         at org.scalajs.logging.Logger.time(Logger.scala:42)
[error]         at org.scalajs.logging.Logger.time$(Logger.scala:40)
[error]         at org.scalajs.sbtplugin.Loggers$SbtLoggerWrapper.time(Loggers.scala:19)
[error]         at org.scalajs.linker.frontend.optimizer.IncOptimizer.update(IncOptimizer.scala:128)
[error]         at org.scalajs.linker.frontend.LinkerFrontendImpl.$anonfun$optimize$1(LinkerFrontendImpl.scala:96)
[error]         at org.scalajs.logging.Logger.time(Logger.scala:42)
[error]         at org.scalajs.logging.Logger.time$(Logger.scala:40)
[error]         at org.scalajs.sbtplugin.Loggers$SbtLoggerWrapper.time(Loggers.scala:19)
[error]         at org.scalajs.linker.frontend.LinkerFrontendImpl.optimize(LinkerFrontendImpl.scala:96)
[error]         at org.scalajs.linker.frontend.LinkerFrontendImpl.$anonfun$link$8(LinkerFrontendImpl.scala:82)
[error]         at scala.concurrent.Future.$anonfun$flatMap$1(Future.scala:307)
[error]         at scala.concurrent.impl.Promise.$anonfun$transformWith$1(Promise.scala:51)
[error]         at scala.concurrent.impl.CallbackRunnable.run(Promise.scala:74)
[error]         at java.base/java.util.concurrent.ForkJoinTask$RunnableExecuteAction.exec(ForkJoinTask.java:1423)
[error]         at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:387)
[error]         at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1312)
[error]         at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1843)
[error]         at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1808)
[error]         at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:188)
[error] Caused by: java.lang.AssertionError: assertion failed: Cannot create a Tree with record type RecordType(List())
[error]         at scala.Predef$.assert(Predef.scala:223)
[error]         at org.scalajs.linker.frontend.optimizer.OptimizerCore$PreTransTree.<init>(OptimizerCore.scala:6684)
[error]         at org.scalajs.linker.frontend.optimizer.OptimizerCore.addPreTransBindings(OptimizerCore.scala:5931)
[error]         at org.scalajs.linker.frontend.optimizer.OptimizerCore.addPreTransBinding(OptimizerCore.scala:5908)
[error]         at org.scalajs.linker.frontend.optimizer.OptimizerCore.$anonfun$withNewLocalDef$3(OptimizerCore.scala:5850)
[error]         at org.scalajs.linker.frontend.optimizer.OptimizerCore.$anonfun$transformBlock$3(OptimizerCore.scala:854)
[error]         at org.scalajs.linker.frontend.optimizer.OptimizerCore.$anonfun$withBinding$1(OptimizerCore.scala:5765)
[error]         at org.scalajs.linker.frontend.optimizer.OptimizerCore.withDedicatedVar$1(OptimizerCore.scala:5849)
[error]         at org.scalajs.linker.frontend.optimizer.OptimizerCore.$anonfun$withNewLocalDef$1(OptimizerCore.scala:5882)
[error]         at scala.util.control.TailCalls$TailRec.result(TailCalls.scala:81)
[error]         at org.scalajs.linker.frontend.optimizer.OptimizerCore.trampoline(OptimizerCore.scala:5980)
[error]         at org.scalajs.linker.frontend.optimizer.OptimizerCore.transformList$1(OptimizerCore.scala:850)
[error]         at org.scalajs.linker.frontend.optimizer.OptimizerCore.transformBlock(OptimizerCore.scala:867)
[error]         at org.scalajs.linker.frontend.optimizer.OptimizerCore.transform(OptimizerCore.scala:332)
[error]         at org.scalajs.linker.frontend.optimizer.OptimizerCore.transformMethodDefBody(OptimizerCore.scala:5514)
[error]         at org.scalajs.linker.frontend.optimizer.OptimizerCore.optimize(OptimizerCore.scala:166)
[error]         at org.scalajs.linker.frontend.optimizer.IncOptimizer$MethodImpl.doProcess(IncOptimizer.scala:1688)
[error]         at org.scalajs.linker.frontend.optimizer.IncOptimizer$MethodImpl.doProcess(IncOptimizer.scala:1626)
[error]         at org.scalajs.linker.frontend.optimizer.IncOptimizer$Processable.process(IncOptimizer.scala:1561)
[error]         at org.scalajs.linker.frontend.optimizer.IncOptimizer.$anonfun$processAllTaggedMethods$3(IncOptimizer.scala:411)
[error]         at org.scalajs.linker.frontend.optimizer.IncOptimizer.$anonfun$processAllTaggedMethods$3$adapted(IncOptimizer.scala:411)
[error]         at scala.collection.parallel.mutable.ParArray$ParArrayIterator.foreach(ParArray.scala:145)
[error]         at scala.collection.parallel.ParIterableLike$Foreach.leaf(ParIterableLike.scala:982)
[error]         at scala.collection.parallel.Task.$anonfun$tryLeaf$1(Tasks.scala:53)
[error]         at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
[error]         at scala.util.control.Breaks$$anon$1.catchBreak(Breaks.scala:67)
[error]         at scala.collection.parallel.Task.tryLeaf(Tasks.scala:56)
[error]         at scala.collection.parallel.Task.tryLeaf$(Tasks.scala:50)
[error]         at scala.collection.parallel.ParIterableLike$Foreach.tryLeaf(ParIterableLike.scala:979)
[error]         at scala.collection.parallel.AdaptiveWorkStealingTasks$WrappedTask.internal(Tasks.scala:170)
[error]         at scala.collection.parallel.AdaptiveWorkStealingTasks$WrappedTask.internal$(Tasks.scala:157)
[error]         at scala.collection.parallel.AdaptiveWorkStealingForkJoinTasks$WrappedTask.internal(Tasks.scala:440)
[error]         at scala.collection.parallel.AdaptiveWorkStealingTasks$WrappedTask.compute(Tasks.scala:150)
[error]         at scala.collection.parallel.AdaptiveWorkStealingTasks$WrappedTask.compute$(Tasks.scala:149)
[error]         at scala.collection.parallel.AdaptiveWorkStealingForkJoinTasks$WrappedTask.compute(Tasks.scala:440)
[error]         at java.base/java.util.concurrent.RecursiveAction.exec(RecursiveAction.java:194)
[error]         at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:387)
[error]         at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.tryRemoveAndExec(ForkJoinPool.java:1351)
[error]         at java.base/java.util.concurrent.ForkJoinTask.awaitDone(ForkJoinTask.java:422)
[error]         at java.base/java.util.concurrent.ForkJoinTask.join(ForkJoinTask.java:651)
[error]         at scala.collection.parallel.ForkJoinTasks$WrappedTask.sync(Tasks.scala:379)
[error]         at scala.collection.parallel.ForkJoinTasks$WrappedTask.sync$(Tasks.scala:379)
[error]         at scala.collection.parallel.AdaptiveWorkStealingForkJoinTasks$WrappedTask.sync(Tasks.scala:440)
[error]         at scala.collection.parallel.ForkJoinTasks.executeAndWaitResult(Tasks.scala:423)
[error]         at scala.collection.parallel.ForkJoinTasks.executeAndWaitResult$(Tasks.scala:416)
[error]         at scala.collection.parallel.ForkJoinTaskSupport.executeAndWaitResult(TaskSupport.scala:60)
[error]         at scala.collection.parallel.ExecutionContextTasks.executeAndWaitResult(Tasks.scala:555)
[error]         at scala.collection.parallel.ExecutionContextTasks.executeAndWaitResult$(Tasks.scala:555)
[error]         at scala.collection.parallel.ExecutionContextTaskSupport.executeAndWaitResult(TaskSupport.scala:84)
[error]         at scala.collection.parallel.ParIterableLike.foreach(ParIterableLike.scala:465)
[error]         at scala.collection.parallel.ParIterableLike.foreach$(ParIterableLike.scala:464)
[error]         at scala.collection.parallel.mutable.ParArray.foreach(ParArray.scala:58)
[error]         at org.scalajs.linker.frontend.optimizer.ParCollOps$.foreach(ParCollOps.scala:50)
[error]         at org.scalajs.linker.frontend.optimizer.ParCollOps$.foreach(ParCollOps.scala:22)
[error]         at org.scalajs.linker.frontend.optimizer.IncOptimizer.processAllTaggedMethods(IncOptimizer.scala:411)
[error]         at org.scalajs.linker.frontend.optimizer.IncOptimizer.$anonfun$update$4(IncOptimizer.scala:128)
[error]         at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
[error]         at org.scalajs.logging.Logger.time(Logger.scala:42)
[error]         at org.scalajs.logging.Logger.time$(Logger.scala:40)
[error]         at org.scalajs.sbtplugin.Loggers$SbtLoggerWrapper.time(Loggers.scala:19)
[error]         at org.scalajs.linker.frontend.optimizer.IncOptimizer.update(IncOptimizer.scala:128)
[error]         at org.scalajs.linker.frontend.LinkerFrontendImpl.$anonfun$optimize$1(LinkerFrontendImpl.scala:96)
[error]         at org.scalajs.logging.Logger.time(Logger.scala:42)
[error]         at org.scalajs.logging.Logger.time$(Logger.scala:40)
[error]         at org.scalajs.sbtplugin.Loggers$SbtLoggerWrapper.time(Loggers.scala:19)
[error]         at org.scalajs.linker.frontend.LinkerFrontendImpl.optimize(LinkerFrontendImpl.scala:96)
[error]         at org.scalajs.linker.frontend.LinkerFrontendImpl.$anonfun$link$8(LinkerFrontendImpl.scala:82)
[error]         at scala.concurrent.Future.$anonfun$flatMap$1(Future.scala:307)
[error]         at scala.concurrent.impl.Promise.$anonfun$transformWith$1(Promise.scala:51)
[error]         at scala.concurrent.impl.CallbackRunnable.run(Promise.scala:74)
[error]         at java.base/java.util.concurrent.ForkJoinTask$RunnableExecuteAction.exec(ForkJoinTask.java:1423)
[error]         at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:387)
[error]         at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1312)
[error]         at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1843)
[error]         at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1808)
[error]         at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:188)
```

</details>

This was discovered in a test in Scala 3, after Scala 3 compiles its own standard library. Scala 3 compiler backend does not try to optimize `Labeled`s coming from pattern match as chains of `If` nodes (yet). That causes `ClassTag.apply` to get a different IR. There was a test with
```scala
object Test {
  val ct = classTag(ClassTag.apply(classOf[String])) // after implicit resolution
}
```
Since `ct` is never read, the assignment ends up in statement position. `classTag` is an inlineable identity function, and so we end up in the situation above.

Here is a minimized reproduction, which crashes in all Scala versions:
```scala
object Test {
  @noinline def testMinimized(): Unit = {
    val instance = makeBug(5)
    val _ = instance
  }

  @inline def makeBug(x: Int): Bug = {
    // Use an explicit `return` to cause a Labeled block and its Return node
    return new Bug(x)
  }
}

@inline
final class Bug(val x: Int)
```

--- a/linker/shared/src/main/scala/org/scalajs/linker/backend/emitter/FunctionEmitter.scala
+++ b/linker/shared/src/main/scala/org/scalajs/linker/backend/emitter/FunctionEmitter.scala
@@ -1666,8 +1666,14 @@ private[emitter] class FunctionEmitter(sjsGen: SJSGen) {
             case Lhs.Return(l) =>
               doReturnToLabel(l)
 
-            case Lhs.ReturnFromFunction | Lhs.Throw =>
-              throw new AssertionError("Cannot return or throw a record value.")
+            case Lhs.ReturnFromFunction =>
+              assert(env.expectedReturnType == VoidType,
+                  "Cannot return a record value from a non-void function")
+              val (newStats, _) = transformBlockStats(elems)
+              js.Block(newStats :+ js.Return(js.Undefined()))
+
+            case Lhs.Throw =>
+              throw new AssertionError("Cannot throw a record value.")
           }
 
         // Control flow constructs

--- a/linker/shared/src/main/scala/org/scalajs/linker/backend/wasmemitter/FunctionEmitter.scala
+++ b/linker/shared/src/main/scala/org/scalajs/linker/backend/wasmemitter/FunctionEmitter.scala
@@ -649,7 +649,7 @@ private class FunctionEmitter private (
       case (NothingType, _) =>
         ()
       case (_, VoidType) =>
-        fb += wa.Drop
+        genDrop(generatedType)
       case (primType: PrimTypeWithRef, _) =>
         // box
         primType match {
@@ -688,6 +688,18 @@ private class FunctionEmitter private (
       case _ =>
         ()
     }
+  }
+
+  /** Drop a value of the given type from the top of the stack. */
+  private def genDrop(tpe: Type): Unit = tpe match {
+    case VoidType =>
+      ()
+    case RecordType(fields) =>
+      for (field <- fields)
+        genDrop(field.tpe)
+    case _ =>
+      // All other types take exactly one slot on the stack
+      fb += wa.Drop
   }
 
   private def genAssign(tree: Assign): Type = {
@@ -3500,8 +3512,7 @@ private class FunctionEmitter private (
             fb += wa.LocalSet(tempLocal)
         } else {
           // Discard this field
-          for (_ <- transformResultType(recordField.tpe))
-            fb += wa.Drop
+          genDrop(recordField.tpe)
         }
       }
 

--- a/linker/shared/src/main/scala/org/scalajs/linker/frontend/optimizer/OptimizerCore.scala
+++ b/linker/shared/src/main/scala/org/scalajs/linker/frontend/optimizer/OptimizerCore.scala
@@ -1572,6 +1572,11 @@ private[optimizer] abstract class OptimizerCore(
    *  By "per translation", we mean in an alternative path through
    *  `tryOrRollback`. It could still be called several times as long as
    *  it is once in the 'try' part and once in the 'fallback' part.
+   *
+   *  !!! #5246 finishTransformStat must never return a Tree whose type is
+   *  a RecordType. Those trees may be put into a PreTransTree, which cannot
+   *  accept them. This method must aggressively get rid of any tree that may
+   *  have a RecordType.
    */
   private def finishTransformStat(stat: PreTransform): Tree = stat match {
     case PreTransBlock(bindingsAndStats, result) =>
@@ -1680,7 +1685,14 @@ private[optimizer] abstract class OptimizerCore(
     }
   }
 
-  /** Keeps only the side effects of a Tree (overapproximation). */
+  /** Keeps only the side effects of a Tree (overapproximation).
+   *
+   *  !!! #5246 keepOnlySideEffects must never return a Tree whose type is
+   *  a RecordType. Those trees may be put into a PreTransTree, which cannot
+   *  accept them. This method must aggressively get rid of any tree that may
+   *  have a RecordType. Last resort: protect it behind a Labeled block with
+   *  type `void`.
+   */
   private def keepOnlySideEffects(stat: Tree): Tree = stat match {
     case _:VarRef | _:Literal | _:SelectStatic =>
       Skip()(stat.pos)
@@ -1716,6 +1728,8 @@ private[optimizer] abstract class OptimizerCore(
         case (Skip(), Skip())     => keepOnlySideEffects(cond)
         case (newThenp, newElsep) => If(cond, newThenp, newElsep)(VoidType)(stat.pos)
       }
+    case Labeled(label, _, body) =>
+      Labeled(label, VoidType, keepOnlySideEffects(body))(stat.pos)
 
     case BinaryOp(op, lhs, rhs) =>
       /* The logic here exceeds the complexity threshold for keeping a copy
@@ -1745,7 +1759,12 @@ private[optimizer] abstract class OptimizerCore(
       Skip()(stat.pos)
 
     case _ =>
-      stat
+      if (stat.tpe.isInstanceOf[RecordType]) {
+        // #5246 Last resort not to return a RecordType'd tree
+        Labeled(freshLabelName(LabelDiscardBase), VoidType, stat)(stat.pos)
+      } else {
+        stat
+      }
   }
 
   private def isNonNegativeIntLiteral(tree: Tree): Boolean = tree match {
@@ -6015,6 +6034,8 @@ private[optimizer] object OptimizerCore {
    *  base.
    */
   private val LocalThisNameForFresh = LocalName("this")
+
+  private val LabelDiscardBase = LabelName("discard")
 
   private val thisOriginalName: OriginalName = OriginalName("this")
 

--- a/test-suite/js/src/test/scala/org/scalajs/testsuite/compiler/OptimizerTest.scala
+++ b/test-suite/js/src/test/scala/org/scalajs/testsuite/compiler/OptimizerTest.scala
@@ -688,6 +688,19 @@ class OptimizerTest {
     }
     assertEquals(1L, z)
   }
+
+  @noinline // just to be sure
+  @Test def inlineClassLabeledReturnDiscardStatementPos_Issue5246(): Unit = {
+    @inline def makeBug(x: Int): Bug5246 = {
+      // Use an explicit `return` to cause a Labeled block and its Return node
+      return new Bug5246(x) // scalastyle:ignore
+    }
+
+    // Assign to val to cause a pretransform
+    val bug = makeBug(5)
+    // but then discard it and leave it in statement position
+    val _ = bug
+  }
 }
 
 object OptimizerTest {
@@ -704,5 +717,8 @@ object OptimizerTest {
     val y = x
     val z = t.y
   }
+
+  @inline
+  final class Bug5246(val x: Int)
 
 }

